crypto/ts/quinTree.ts

```diff
@@ -0,0 +1,377 @@
+    while (numLeaves !== this.arity ** subDepth && subDepth < this.depth) {
+      subDepth += 1;
+    }
+
+    const subTree = new IncrementalQuinTree(subDepth, this.zeroValue, this.arity, this.hashFunc);
+
+    for (let i = startIndex; i < endIndex; i += 1) {
+      subTree.insert(this.getNode(i));
+    }
+
+    const fullPath = this.genProof(startIndex);
+    fullPath.pathIndices = fullPath.pathIndices.slice(subDepth, this.depth);
+    fullPath.pathElements = fullPath.pathElements.slice(subDepth, this.depth);
+    fullPath.leaf = subTree.root;
+
+    return fullPath;
+  }
+
+  /**
+   * Verify a proof
+   * @param proof The proof to verify
+   * @returns Wether the proof is valid
+   */
+  verifyProof = (proof: IMerkleProof): boolean => {
+    const { pathElements, leaf, root, pathIndices } = proof;
+
+    // Hash the first level
+    const firstLevel: bigint[] = pathElements[0].map(BigInt);
+
+    firstLevel.splice(Number(pathIndices[0]), 0, leaf);
+
+    let currentLevelHash: bigint = this.hashFunc(firstLevel);
+
+    // Verify the proof
+    for (let i = 1; i < pathElements.length; i += 1) {
+      const level: bigint[] = pathElements[i].map(BigInt);
+      level.splice(Number(pathIndices[i]), 0, currentLevelHash);
+
+      currentLevelHash = this.hashFunc(level);
+    }
+
+    // the path is valid if the root matches the calculated root
+    return currentLevelHash === root;
+  };
+
+  /**
+   * Calculate the indices of the parent
+   * @param index The index of the leaf
+   * @returns The indices of the parent
+   */
+  calcParentIndices(index: number): number[] {
+    // can only calculate the parent for leaves not subroots
+    if (index >= this.capacity) {
+      throw new Error(`Index ${index} is out of bounds. Can only get parents of leaves`);
+    }
+
+    const indices = new Array<number>(this.depth);
+    let r = index;
+    let levelCapacity = 0;
+
+    for (let i = 0; i < this.depth; i += 1) {
+      levelCapacity += this.arity ** (this.depth - i);
+      r = Math.floor(r / this.arity);
+      indices.push(levelCapacity + r);
+    }
+
+    return indices;
+  }
+
+  /**
+   * Calculate the indices of the children of a node
+   * @param index The index of the node
+   * @returns The indices of the children
+   */
+  calcChildIndices(index: number): number[] {
+    // cannot get the children of a leaf
+    if (index < this.capacity) {
+      throw new Error(`Index ${index} is out of bounds. Can only get children of subroots`);
+    }
+
+    // find the level
+    let level = 0;
+    let r = this.arity ** level;
+    do {
+      level += 1;
+      r += this.arity ** level;
+    } while (index >= r);
+
+    const start = (index - this.arity ** level) * this.arity;
+    const indices = Array<number>(this.arity)
+      .fill(0)
+      .map((_, i) => start + i);
+
+    return indices;
+  }
+
+  /**
+   * Get a node at a given index
+   * @param index The index of the node
+   * @returns The node
+   */
+  getNode(index: number): Leaf {
```

Comment:
should we also check if it's **out of bounds** here?

Reply:
don't think there's any need here, as checks are done where we call it from and it's never used directly